src/lib.rs

```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
+        }
 
-        let e = match options.epsilon {
-            None => std::time::Duration::from_secs(0),
-            Some(e) => e,
-        };
+        if options.issuer == Required && self.issuer.is_none() {
+            missing_claims.push("iss");
+        }
 
-        if self.expiry.is_some() && !Self::is_after(*self.expiry.unwrap(), now, e)? {
-            Err(ValidationError::TemporalError("Token expired".to_string()))?;
+        if options.subject == Required && self.subject.is_none() {
+            missing_claims.push("sub");
         }
 
-        if self.issued_at.is_some() && !Self::is_before(*self.issued_at.unwrap(), now, e)? {
-            Err(ValidationError::TemporalError("Token issued in the future".to_string()))?;
+        if options.id == Required && self.id.is_none() {
+            missing_claims.push("jti");
         }
 
-        if self.not_before.is_some() && !Self::is_before(*self.not_before.unwrap(), now, e)? {
-            Err(ValidationError::TemporalError("Token not valid yet".to_string()))?;
+        if missing_claims.len() == 0 {
+            Ok(())
+        } else {
+            Err(ValidationError::MissingRequiredClaims(
+                missing_claims.into_iter().map(|v| v.into()).collect(),
+            ))
         }
+    }
 
-        Ok(())
+    ///Validates that if the token has an `exp` claim, it has not passed.
+    pub fn validate_exp(&self, validation: Validation<TemporalOptions>) -> Result<(), ValidationError> {
+        match validation {
+            Validation::Ignored => Ok(()),
+            Validation::Validate(temporal_options) => {
+                let now = temporal_options.now.unwrap_or(Utc::now());
+
+                match self.expiry {
+                    Some(Timestamp(expiry)) if now - expiry > temporal_options.epsilon => {
+                        Err(ValidationError::Expired(now - expiry))
+                    }
+                    _ => Ok(()),
+                }
+            }
+        }
     }
 
-    /// Check `a` is after `b` within a tolerated duration of `e`, where `e` is unsigned: a - b >= -e
-    fn is_after<Tz, Tz2>(a: DateTime<Tz>, b: DateTime<Tz2>, e: std::time::Duration) -> Result<bool, ValidationError>
-    where
-        Tz: chrono::offset::TimeZone,
-        Tz2: chrono::offset::TimeZone,
-    {
-        // FIXME: `chrono::Duration` is a re-export of `time::Duration` and this returns has an error of type
-        // `time::OutOfRangeError`. We don't want to put `time` as a dependent crate just to `impl From` for this...
-        // So I am just going to `map_err`.
-        use std::error::Error;
+    ///Validates that if the token has an `nbf` claim, it has passed.
+    pub fn validate_nbf(&self, validation: Validation<TemporalOptions>) -> Result<(), ValidationError> {
+        match validation {
+            Validation::Ignored => Ok(()),
+            Validation::Validate(temporal_options) => {
+                let now = temporal_options.now.unwrap_or(Utc::now());
```

Comment:
Use [`unwrap_or_else`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_else) for lazy evaluation.